pkg/controller/statusmanager/pod_status.go

```diff
@@ -61,17 +64,135 @@ type deploymentState struct {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+func SetNodeCondition(conditions *[]corev1.NodeCondition, newCondition corev1.NodeCondition) {
+	if conditions == nil {
+		conditions = &[]corev1.NodeCondition{}
+	}
+	existingCondition := FindStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		*conditions = append(*conditions, newCondition)
+		return
+	}
+
+	if existingCondition.Status != newCondition.Status {
+		existingCondition.Status = newCondition.Status
+		existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+	}
+
+	existingCondition.Reason = newCondition.Reason
+	existingCondition.Message = newCondition.Message
+}
+
+func (status *StatusManager) setNodeNetworkUnavailable(nodeName string, ready bool, reason string, message string) {
+	if ready == true {
+		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to true for node %s", nodeName))
+	} else {
+		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to false for node %s", nodeName))
+	}
+	node := &corev1.Node{}
+	err := status.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Failed to get node %s", nodeName))
+		return
+	}
+	networkUnavailableCondition := corev1.ConditionFalse
+	if ready == false {
+		networkUnavailableCondition = corev1.ConditionTrue
+		reason += "NotReady"
+	} else {
+		reason += "Ready"
+		message = "NSX node agent is running"
+	}
+	condition := corev1.NodeCondition{
+		Type:               "NetworkUnavailable",
+		Status:             networkUnavailableCondition,
+		LastHeartbeatTime:  metav1.Now(),
+		LastTransitionTime: metav1.Now(),
+		Reason:             reason,
+		Message:            message,
+	}
+	SetNodeCondition(&node.Status.Conditions, condition)
+	err = status.client.Status().Update(context.TODO(), node)
+	if err != nil {
+		log.Error(err, "Failed to update node condition")
+	} else {
+		log.Info("Updated node condition")
+	}
+}
+
+// Get the pod status from API server
+func (status *StatusManager) SetFromPodsForNodeCondition() {
+	status.Lock()
+	defer status.Unlock()
+
+	pods := &corev1.PodList{}
+	err := status.client.List(context.TODO(), pods, client.MatchingLabels{"component": operatortypes.NsxNodeAgentContainerName})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Error getting %s for node condition", operatortypes.NsxNodeAgentContainerName))
+	} else if len(pods.Items) == 0 {
+		log.Info("nsx-node-agent not found for node condition")
+		nodes := corev1.NodeList{}
+		errNode := status.client.List(context.TODO(), &nodes)
+		if errNode != nil {
+			log.Error(errNode, "Failed to get nodes for condition updating")
+			return
+		}
+		for _, node := range nodes.Items {
+			status.setNodeNetworkUnavailable(node.ObjectMeta.Name, false, "NSXNodeAgent", "Waiting for nsx-node-agent to be created")
```

Comment:
Please check node existing status to avoid updating the same condition many times, you can add it in a following patch

Reply:
hmm... I saw in status_manager.go, the definition of set() also directly invokes client.Status().Update(). Not sure if the Update() itself can avoid the dup updating, I'm adding a TODO to confirm it.